formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufEncoding.kt

```diff
@@ -198,18 +201,24 @@ private class OneOfClassEncoder(
     proto: ProtoBuf,
     parentTag: ProtoDesc,
     private val parentWriter: ProtobufWriter,
-    private val stream: ByteArrayOutput = ByteArrayOutput(),
     descriptor: SerialDescriptor
 ) : ProtobufEncoder(proto, parentWriter, descriptor) {
```

Comment:
> You want to (indirectly) implement CompositeEncoder as well as Encoder. You require beginStructure to be called (returning this). Then for the structure you might record the string type written as element 0 (in this case you could ignore it, or keep it for checking/diagnostics), and the actual value with its serializer as element 1 (you can here use the descriptor of the serializer to get the ProtoNum) 

That was my very first implementation of encoding, but I soon found that I was somehow hijacking a generic polymorphic serializer, by assuming that the serializer would write a string first which I can safely ignore and the actual content to accept. 

The encoder will nolonger be stateless since function `encodeIntElement` with `index` parameter is final in base `ProtobufTaggedEncoder` so I can only know whether the serializer is writing the discriminator with a boolean property `discriminatorEcoded` set to be true in `encodeTaggedString`. 

But I realize that I can get the index in `tag` of `encodeTaggedString` by writing this reply, and so far is OK for me.

When comes to decoding, I need to mock the behaviour how a polymorphic serializer read contents: Feeds it with a string of class discriminator, then the actual content. To get the discriminator, I need to find the serialName of actual serializer, which means that I need to resolve the actual serializer from polymorphic. But since I am getting the actual serializer, why not just use it for decoding?

Do you have any suggesting for getting a proper `serialName` for polymorphic serializer?

Reply:
I managed to rewrite the encoding, it is much more elegent than what I thought before, thanks for your review.

But now I am still struggling in the decoding.